controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/DeleteReaderGroupTask.java

```diff
@@ -0,0 +1,85 @@
+                                 final ScheduledExecutorService executor) {
+        Preconditions.checkNotNull(streamMetaStore);
+        Preconditions.checkNotNull(streamMetaTasks);
+        Preconditions.checkNotNull(executor);
+        this.streamMetadataStore = streamMetaStore;
+        this.streamMetadataTasks = streamMetaTasks;
+        this.executor = executor;
+    }
+
+    @Override
+    public CompletableFuture<Void> execute(final DeleteReaderGroupEvent request) {
+      String scope = request.getScope();
+      String readerGroup = request.getRgName();
+      long requestId = request.getRequestId();
+      UUID readerGroupId = request.getReaderGroupId();
+      final RGOperationContext context = streamMetadataStore.createRGContext(scope, readerGroup);
+      return streamMetadataStore.getReaderGroupId(scope, readerGroup, context, executor)
+              .thenCompose(id -> {
+               if (!id.equals(readerGroupId)) {
+                      log.warn("Skipping processing of Reader Group delete request {} as UUIDs did not match.", requestId);
+                      return CompletableFuture.completedFuture(null);
+               }
+               return streamMetadataStore.getReaderGroupConfigRecord(scope, readerGroup, context, executor)
+                          .thenCompose(configRecord -> {
+                          if (!ReaderGroupConfig.StreamDataRetention.values()[configRecord.getObject().getRetentionTypeOrdinal()]
+                              .equals(ReaderGroupConfig.StreamDataRetention.NONE)) {
+                              String scopedRGName = NameUtils.getScopedReaderGroupName(scope, readerGroup);
+                              // update Stream metadata tables, if RG is a Subscriber
+                              Iterator<String> streamIter = configRecord.getObject().getStartingStreamCuts().keySet().iterator();
+                              return Futures.loop(() -> streamIter.hasNext(), () -> {
+                                     Stream stream = Stream.of(streamIter.next());
+                                     return streamMetadataStore.deleteSubscriber(stream.getScope(),
+                                            stream.getStreamName(), scopedRGName, configRecord.getObject().getGeneration(), null, executor);
+                                 }, executor);
+                              }
+                              return CompletableFuture.completedFuture(null);
+                          })
+                          .thenCompose(v -> streamMetadataTasks.sealStream(scope, NameUtils.getStreamForReaderGroup(readerGroup), null))
+                                  .thenCompose(v -> streamMetadataTasks.deleteStream(scope,
+                                  NameUtils.getStreamForReaderGroup(readerGroup), null)
+                                  .thenCompose(v1 -> streamMetadataStore.deleteReaderGroup(scope, readerGroup, context, executor)));
```

Comment:
should we pass the rg UUID to delete RG method and to a conditonal delete just to be extra safe? i know you check the rg id before processing the event. and the create is also linearized on this processor so they do not race per se. but still a conditional delete may be a safer thing to do. perhaps open a github issue to address this. thanks

Reply:
fixed